[#1844] Support specific Markdown comment formats for @@author

parth-io opened the pull request:
Fixes #1844

## Proposed commit message
```
There are many comment formats for Markdown files, and not all comment
formats are supported by the @@author tags. Users of RepoSense may find
behavior of @@author tags in Markdown comments inconsistent.

Let's add support for the other Markdown comment formats for the
@@author tag.
```

## Other information
There are 3 main Markdown formats:
```
[filler text here]: # (comment)
<!-- comment -->
<!--- comment --->
```
The triple-dashed comment seems to be the most widely supported, and GFMD supports all 3. There are other formats specific to other variants, which have not been accounted for.

Also, it might be a simpler design to store the comment patterns as a static variable under a class `MarkdownFileInfo` that inherits from `FileInfo`. `AnnotationAnalyzer` feels unnecessarily complex, with methods such as `getCommentTypeIndex()` that perform additional work that can be avoided with a simpler design.

This can help if it is decided to associate different comment formats with different FileTypes in the future (seems unlikely/unneeded). Each comment pattern may need a unique id in that scenario, which was not implemented in this PR.

chan-j-d reviewed (commented):
Thanks for the PR and interesting suggestion. But I wanted to ask what are the disadvantages of just including the regex for the new comment type instead of separating it into 2 cases.

parth-io commented:
The first consideration was that of performance. I did not run any profile tests, but adding 2 new comment patterns to the existing 6 comment patterns seemed likely to slow down the code, given that currently the code inspects every line in the project for all comment patterns, regardless of file type. Thus, limiting the two new comment patterns to markdown files seemed to be the way to limit any impact on performance.

I also intended to refactor the code for this reason and limit specific comment patterns to their respective file types, but I did not do so as it might have been out of scope for this PR.

A minor reason was that Markdown also serves as documentation, so in the unlikely event that anyone writes a comment with an author tag in markdown files as part of documentation for code in other files, RepoSense will not incorrectly flag it as marking author contributions.

chan-j-d commented:
To the first point, performance is generally not a concern as there are other much more problematic bottlenecks for reposense I believe. Additionally, I think checking filetypes incurs around the same penalty anyway.

My preference would be to just add the new comment patterns to the existing ones. I guess my main issue is that I think specifying comment pattern per file type brings about quite a bit of maintenance burden without too much payoff at the moment but might be something we can consider in the future if it becomes a huge issue (consider creating a new issue for this). 

Also the reason you state is somewhat strange since it seems like the only reason the author tag would appear in documentation as part of a comment is for reposense to detect it since it.

parth-io commented:
> checking filetypes incurs around the same penalty anyway.

Can I check this reasoning: I was thinking that the method `getCommentTypeIndex` in `AnnotationAnalyzer.java` has a for-loop that runs for every line, so it might slow down RepoSense. We can instead do a single file check at the start in `aggregateAnnotationAuthorInfo`.

I understand that this is not a performance bottleneck, I will rewrite the PR to simplify it then.

Admittedly, the second reason is a bit unusual. I was thinking that if someone added sample Python code to a code block in the markdown file and left the author tags from the original Python file in, RepoSense should not detect it as marking author contributions. But this might be over-engineering, because I cannot think why anyone would leave the `@@author` tag in the sample code.

chan-j-d commented:
On the first point, I think you're right. My bad as it's been awhile since I've taken a close look at the original code. It seems that there will be some marginal improvement made in this change. Regardless, I think it's better to leave it as it is for simplicity.

parth-io commented:
Hi @chan-j-d, I have updated the PR to be simpler. Do let me know if I am missing anything else to add (unit tests, documentation, etc.)

I also updated the proposed commit message.

chan-j-d reviewed (changes requested):
Some very minor comments/changes. The rest looks good!

parth-io commented on docs/ug/usingAuthorTags.md:
```diff
@@ -32,14 +32,21 @@ The `@@author` tags should be enclosed within a comment, using the comment synta
+<ul>
+    <li>&lt!-- @@author authorName --&gt</li>
+    <li>&lt!--- @@author authorName ---&gt</li>
+    <li>[//]: # (@@author authorName)</li>
```
Fixed!

parth-io commented on src/main/java/reposense/authorship/analyzer/AnnotatorAnalyzer.java:
```diff
@@ -32,20 +32,24 @@ public class AnnotatorAnalyzer {
      * Overrides the authorship information in {@code fileInfo} based on annotations given on the file.
      *
-     * @param fileInfo FileInfo to be further analyzed with author annotations.
+     * @param fileInfo     FileInfo to be further analyzed with author annotations.
```
Fixed

chan-j-d commented:
@parth-io some of the tests are failing. Might want to go and take a look

parth-io commented:
Hi @chan-j-d, fixed the failing tests.

chan-j-d reviewed (approved):
LGTM!

dcshzj commented:
@parth-io Sorry, the tests are still failing, are you able to look into it again? Let us know if you have any questions!

parth-io commented:
Hi @dcshzj yep, I'm investigating - it's the MacOS unit tests that are failing, for unrelated code. I'll see what I can figure out first and let you know

parth-io commented:
@dcshzj I'm not exactly sure what's causing the issue as I don't have a Mac and I can't spin up a Mac VM now. However, there are other open PRs which are failing the MacOS unit tests despite passing the Ubuntu and Windows unit tests:
1. PR #1852 - https://github.com/reposense/RepoSense/actions/runs/3738958992
2. PR #1850 - https://github.com/reposense/RepoSense/actions/runs/3733277182/jobs/6380241650

The last successful CI run that passed the MacOS unit tests was https://github.com/reposense/RepoSense/actions/runs/3716783099, on 17 Dec, for commit b2892daa402721fe0f051ca9b3ffe9e455adf546

Commit 0b7a7aec might be causing the issue then, but it passed the tests then

dcshzj commented:
Interesting, I ran the tests on my Mac and the tests are all passing. We might have to further investigate more specifically on the environment that the tests are running on (i.e. single threaded, UTC timezone, etc) and why they are causing issues for the tests.

dcshzj commented:
The runner on GitHub Actions is using Git version 2.39.0 which broke the tests. The tests were passing on version 2.38.1 on my Mac originally.

dcshzj reviewed (approved):
LGTM! I think this PR is quite well done, thanks for contributing!
